extra/tensors/tensors.factor

```diff
@@ -0,0 +1,257 @@
+! Copyright (C) 2019 HMC Clinic.
+! See http://factorcode.org/license.txt for BSD license.
+USING: accessors alien.c-types alien.data arrays
+concurrency.combinators grouping kernel locals math.functions
+math.ranges math multi-methods quotations sequences sequences.private
+specialized-arrays tensors.tensor-slice typed ;
+QUALIFIED-WITH: alien.c-types c
+SPECIALIZED-ARRAY: c:float
+IN: tensors
+
+! Tensor class definition
+TUPLE: tensor
+    { shape array }
+    { vec float-array } ;
+
+! Errors
+ERROR: shape-mismatch-error shape1 shape2 ;
+
+<PRIVATE
+
+! Construct a tensor of zeros
+: <tensor> ( shape seq -- tensor )
+    tensor boa ;
+
+: >float-array ( seq -- float-array )
+    [ ] float-array{ } map-as ;
+
+:: repetition ( shape const -- tensor )
+    ! Push the shape
+    shape
+    ! Push the float array
+    shape product const <repetition>
+    >float-array
+    ! Create the tensor
+    <tensor> ;
+
+PRIVATE>
+
+! Construct a tensor of zeros
+: zeros ( shape -- tensor )
+    0 repetition ;
+
+! Construct a tensor of ones
+: ones ( shape -- tensor )
+    1 repetition ;
+
+! Construct a one-dimensional tensor with values start, start+step,
+! ..., stop (inclusive)
+: arange ( a b step -- tensor )
+    <range> [ length 1array ] keep [ ] float-array{ } map-as <tensor> ;
+
+! Construct a tensors with vec { 0 1 2 ... } and reshape to the desired shape
+: naturals ( shape -- tensor )
+    [ ] [ product [0,b) >float-array ] bi <tensor> ;
+
+! Reshape the tensor to conform to the new shape
+: reshape ( seq tensor -- tensor )
```

Comment:
You might change the stack effect to be ``( shape tensor -- tensor )``, and maybe think about re-ordering stack effects, and extracting the shape checking out:

```
: check-shapes ( shape1 shape2 -- shape1 shape2 )
    2dup [ product ] bi@ = [ shape-mismatch-error ] unless ;

: reshape ( tensor shape -- tensor )
    '[ _ check-shapes nip ] change-shape ;
```

Reply:
I updated this, but the reshape implementation is not quite what you suggested. Let me know if there are any further changes.